### ISSUE TITLE
Service discovery on header other than authority

### ISSUE THREAD

**Issue by grampelberg** [linkerd/linkerd2#1998]
### What problem are you trying to solve?

It is valuable to inject an ingress controller with LD2. As ingress, the `host` header tends to be the external host and most controllers just route directly back to the IP address of an endpoint. Unfortunately, this makes it so that Linkerd cannot discover the destination (as that works off the `host` header). 

### How should the problem be solved?

Add a new header that can be used for service discovery. My proposal is `X-Forwarded-To`.

### Concerns

What are the security implications of this? Would we need to strip the `X-Forwarded-To` header on the server side and only use it on the client side?

**Comment by seanmonstar** [comment]
(Meta note: this may not yet be a "good first issue", since we don't yet have a design specced out.)

I suppose we can't rely on the Destination service to help us, because the `host` will look like an external name, and so it'd just do a DNS lookup and then probably loop on itself.

**Comment by grampelberg** [comment]
@seanmonstar I don't understand the implications. Why can't we use the same, existing service discovery logic and just pick from a different header?

**Comment by seanmonstar** [comment]
Sorry, we could, I was just describing a bit more what would happen *now*, as of 2.1.

Well... actually, let me see. If a proxy is installed on the ingress controller, it will be forwarding requests from the outside world to the wrapped controller. That'd mean it was *inbound*, so we'd just pass the request to the original controller, and not actually do any service discovery.

If we wanted to completely remove the existing ingress controller, then we'd need to do something, ya.

**Comment by grampelberg** [comment]
Lemme try explaining the process as I see it:

1. Request comes in to pod, `Host` header is `example.com`
1. LD2 proxy receives request, passes on to nginx
1. Nginx sees `Host: example.com`, matches with its config
1. Nginx sends a request to the actual service pod, let's say `10.1.1.5`. `Host: example.com` is still there.
1. LD2 sees `Host: example.com` and instead of passing the request back to `10.1.1.5`, it issues a DNS request for `example.com`
1. Sadness

I'm suggesting that Nginx add a new header for the request from Nginx to the actual service pod. Then, LD2 will see `X-Forwarded-To` (for example), do service discovery and figure out where the request should finally end at.

**Comment by seanmonstar** [comment]
Yep, I see what you mean.

1. Should we even be wrapping an ingress controller, instead of just *being* the ingress controller?
2. If it is always forwarding to an IP address, we could have see about allowing that to be configured directly during `inject`, and not need to check another header. Or does Nginx often forward to an internal name?

**Comment by hawkw** [comment]
@grampelberg Am I understanding correctly that the proposed change is for Linkerd to fall back to another header (let's say `X-Forwarded-For` for brevity's sake) when a lookup for the `Host` header corresponds to the pod's IP? Or is the intention for `X-Forwarded-To` to take _priority_ over the host header when it is present, so the DNS lookup for the host header is avoided?

**Comment by olix0r** [comment]
`X-*` headers should be considered deprecated, in general... Is this some psuedo-standard?

**Comment by seanmonstar** [comment]
[`X-Forwarded-For`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For) is a pretty common header that reverse proxies add (or the standardized version, `Forwarded`).

**Comment by olix0r** [comment]
The suggested value is `X-Forward-To` (not XFF).

**Comment by seanmonstar** [comment]
Woops, derp moment!

**Comment by grampelberg** [comment]
I threw `X-Forward-To` as a starter. I'd rather that someone who knows headers better than I pick what it should be =)

@hawkw I think the header should take priority, but only for incoming localhost connections. It is unlikely that the `Host` domain would ever resolve to the pod's IP.

**Comment by olix0r** [comment]
This should probably be a `l5d-*` header, as I don't think we want this to be semantically meaningful outside of linkerd, do we? I have some additional ideas about how and when to allow headers to influence routing which I'm in the process of writing up...

**Comment by grampelberg** [comment]
+1 from me, I don't see any reason for it to be semantically meaningful outside of linkerd.

**Comment by hawkw** [comment]
Yeah, that seems correct to me as well. @olix0r I'm guessing any work on this ticket should wait until you've written up those thoughts?

**Comment by olix0r** [comment]
@hawkw let's get as specific as we can on the plan for this change, and then we can use my doc as a way to validate that plan.

**Comment by hawkw** [comment]
okay, so as I understand it, the proposed change here is:
* bikeshed what the header we want to route on is (for the remainder of this comment I'll call it `l5d-forward-to`)
* add a function in `src/app/mod.rs` like `http_request_l5d_forward_to_addr`, which if that header is present and its' value is a valid authority, returns an `Addr` for that authority,
* change https://github.com/linkerd/linkerd2-proxy/blob/82524e4a1f5d846fbcae259c92253543d067d6c6/src/app/main.rs#L412-L419
to something like this:
  ```rust
                    .push(router::layer(|req: &http::Request<_>| {
                        let addr = super::http_request_l5d_forward_to_addr(req)
                            .or_else(|_| super::http_request_authority_addr(req))
                            .or_else(|_| super::http_request_host_addr(req))
                            .or_else(|_| super::http_request_orig_dst_addr(req))
                            .ok();
                        debug!("outbound addr={:?}", addr);
                        addr
                    }))
  ```
  so that the outbound router will preferentially choose the `l5d-forward-to` authority if one is present, and
* document this so that nginx-ingress users know to set that header

@olix0r, @grampelberg, and @seanmonstar, does that seem in line with your understanding here?

**Comment by grampelberg** [comment]
> consider having the proxy strip that header from the forwarded request, once it has extracted the authority from it

Naively, I assumed that it would be nice to leave the header in (primarily for debugging). I don't know what the specific implications are however.

> document this so that nginx-ingress users know to set that header

I can take this one, it'll be for traefik and the envoy based ingress controllers as well.

👍 from my perspective.

**Comment by hawkw** [comment]
@grampelberg 
> > consider having the proxy strip that header from the forwarded request, once it has extracted the authority from it
>
> Naively, I assumed that it would be nice to leave the header in (primarily for debugging). I don't know what the specific implications are however.

If only the outbound side of a proxy routes based on this header, the proxy receiving an inbound request from the ingress-controller's proxy will still route it to that proxy's pod correctly. 

However, I _can_ see an issue occurring if a service with the proxy injected receives the request and then propagates the header (such as by forwarding the request to another service). In such a case, that service's proxy's outbound router will receive the request and route it back to itself, potentially creating a loop. Hopefully that explanation makes sense?

I don't see any _obvious_ security concerns resulting from not stripping this header, but I may be missing something on that front as well.

**Comment by grampelberg** [comment]
That was my thinking too. TBH I don't have a super solid reason to leave it in 🤷‍♀️

**Comment by hawkw** [comment]
I'm not sure if a service propagating that header would be a particularly common case, but IMO avoiding the potential for a weird failure outweighs the benefits of propagating the header for debugging purposes...

**Comment by olix0r** [comment]
> okay, so as I understand it, the proposed change here is:
> 
> * bikeshed what the header we want to route on is (for the remainder of this comment I'll call it `l5d-forward-to`)

I nominate `l5d-dst-override`. This is in-line with the outbound headers added by the proxy. The only time we talk about "forwarding" in the proxy, currently, is in the context of TCP proxying.

> * consider having the proxy strip that header from the forwarded request, once it has extracted the authority from it, and

The inbound and outbound proxies should always strip this header, since this header can potentially influence routing later if it's passed through unintentionally. The `l5d-dst-*` headers we already add should be suitable for debugging, I think.

Should canonicalization be performed when this header is added? or should this header be assumed to be a FQDN?

**Comment by grampelberg** [comment]
@olix0r what are we doing with `Host` today?

**Comment by olix0r** [comment]
@grampelberg host is canonicalized because the application may perform DNS resolution first. In this case, we're not expecting the header-specifier to have previously done DNS resolution. That said, it's probably best to canonicalize so that we can accept the broadest set of inputs.

**Comment by grampelberg** [comment]
From a UX side of things, we're going to need to document this and explain what you put in. So, we can require that it is a FQDN. Obviously it would be nice to have looser restrictions but that's a P2 in my mind.

The P1 for this would be debugging. There needs to be a way to see whether this is working or failing and what to do with it.

**Comment by olix0r** [comment]
@grampelberg it's easier to do canonicalization, so let's just do that.

@grampelberg What debugging workflow are you imagining?

**Comment by grampelberg** [comment]
I need to spend some more time thinking about the meta issue. The problem that should be solved is a way to reason about why your traffic is not getting where you want it to. Specifically, the folks who say "I added LD to my app and now everything is timing out!"

I'm envisioning a `tap` style stream that shows the relevant routing information (host, destination ip, whether we got it from the control plane and whether we were able to make the connection).
